python/shape.cpp

```diff
@@ -40,12 +40,33 @@ template <class T> void bind_concatenate(py::module &m) {
 template <class T> void bind_reshape(pybind11::module &mod) {
   mod.def(
       "reshape",
-      [](const T &self, const std::vector<Dim> &labels,
-         const py::tuple &shape) {
-        Dimensions dims(labels, shape.cast<std::vector<scipp::index>>());
-        return reshape(self, dims);
+      [](const T &self, const py::dict &dims) {
+        Dimensions new_dims;
+        for (const auto &item : dims)
+          new_dims.addInner(item.first.cast<Dim>(),
+                            item.second.cast<scipp::index>());
+        return reshape(self, new_dims);
       },
-      py::arg("x"), py::arg("dims"), py::arg("shape"));
+      py::arg("x"), py::arg("dims"));
+}
+
+template <class T> void bind_reshape_data_array(pybind11::module &mod) {
+  mod.def(
+      "reshape",
+      [](const T &self, const Dim dim, const py::dict &to_dims) {
+        Dimensions new_dims;
+        for (const auto &item : to_dims)
+          new_dims.addInner(item.first.cast<Dim>(),
+                            item.second.cast<scipp::index>());
+        return reshape(self, dim, new_dims);
+      },
+      py::arg("x"), py::arg("dim"), py::arg("to_dims"));
+  mod.def(
+      "reshape",
```

Comment:
i have no strong feelings about this. I thought `stack` and `unstack` in `xarray` were actually slightly different (e.g. you can unstack/flatten a dim that has not already been stacked), and users would then not have to learn a new terminology, since we are basically just doing a reshape.

Reply:
The problem I see is the if we use `reshape` we have multiple conflicting signatures. This is quite normal in a C++ API with overloads, but it is odd that only certain combinations of args are supported. We had a similar dicussion a couple month ago which led us to split the overloaded `Variable` constructor into `sc.scalar`, `sc.array`, and `sc.zeros`.